VSCode Logging

mj023 opened the pull request:
Add additional logging for the VSCode Extension.

#### Changes

The logging information will be in JSON-Format and is sent to a port that is specified by the user, by setting an environment variable. If the environment variable is not set, no logging will happen. The communication happens in a new thread so that pytask is not blocked. Deamon Threads are not used, because we do not want the threads to be killed when the main pytask program ends, as this can result in them not being able to send the info if pytask finishes very quickly.

#### Info the extension needs
- Collection: The ExitCode of the Collection and a list of the collected tasks with the name and path of the task.
Example: ```{exitcode: "OK", tasks: [ {name: "task1", path: "C:\Documents\task_task1.py"}, ...]}```

- Execution: Every time a task finishes, the name, the outcome and if the task failed also the execution info eg. the error message.
Example: ```{name: "task1", outcome: "FAIL", exc_info: "Rendered String of Error message"}```

mj023 commented:
I added the logging code for the task execution to the hook in live.py, but this seems a bit odd considering the project structure. I could move it to a new implementation of the hook in logging.py that uses ```tryfirst``` or implement the hook a second time in execute.py, also with ```tryfirst```. What do you think, @tobiasraabe ?

tobiasraabe commented on src/_pytask/live.py:
```diff
@@ -168,6 +171,27 @@ def pytask_execute_task_log_start(self, task: PTask) -> bool:
     def pytask_execute_task_log_end(self, report: ExecutionReport) -> bool:
         """Mark a task as being finished and update outcome."""
+        if report.outcome == TaskOutcome.FAIL and report.exc_info is not None:
+            with console.capture() as capture:
```
Check out `render_to_string` from `console.py`

mj023 commented:
- Okay, I'll add a new module then. 
- The first solution to the guarding problem I can think of, would be to add this as an option in the pytask config that I can activate with a cli argument. But I will see if I can check whether the process that calls the script is VS Code or if I can maybe set some environment variable before running pytask.
- I'll try to change it to the standard urllib, requests was just easier to use.

mj023 commented:
I think everything should be ready now, but I'm having a little trouble writing the tests. I have moved the code to a new vscode.py module, replaced the requests library with the standard urllib and added a guard so the code is only executed if a specific environment variable is set. I then added two new tests, so that almost everything I added is covered, but I'm not sure if I did that correctly. Furthermore the tests on ubuntu 3.12 now failed and I can not figure out why.
